### ISSUE TITLE
vcons doesn't work since 1.16.x beta

### ISSUE THREAD

**Issue by mikrosk** [freemint/freemint#118]
Another user report - [virtual consoles](http://joy.sophics.cz/www/ttyvdeva.lzh) worked in FreeMiNT up to this version. I don't see anything specific in git log which would remove support for it (it always has been a kind of hack) and I admit, I have never even tried it (I'm from the "new generation" of freemint users :) ) but there are users who do remember and don't use current snapshots because of it.

**Comment by ps68060** [comment]
I remember the discussion on the old mint list. This is the nearest I could find.

https://mikro.naprvyraz.sk/mint/200806/msg00020.html

**Comment by paulwratt** [comment]
I think there was also a vcons deamon in a sparemint rpm on the old atariforge (sparemint) server, I remember reading the developer notes when I was doing AFROS-update (2009?), but I could not figure a good way to get the key combos to work because of linux local host

**Comment by mikrosk** [comment]
The last working commit with vcons (supplied by Swe, used the `vcons.original` on my TT) is 63ac8e3b. 50a28d07 results in a yellow screen when switching to another console and while the following one (67e3237e) seemingly fixes it, it is easy to freeze the whole system (my test case was: login on tty1, run `top`, login on tty2, type `ls` + TAB = freeze).

Note: none of these commits seem to work on Aranym (latest build + latest EmuTOS), the console seems to be frozen. Most likely is not frozen but some screen switching is silently ignored.

**Comment by swe-yc** [comment]
> The last working commit with vcons (supplied by Swe, used the `vcons.original` on my TT) is [63ac8e3](https://github.com/freemint/freemint/commit/63ac8e3bbb70154fe95e73998cc6309f42e22657). [50a28d0](https://github.com/freemint/freemint/commit/50a28d070efc771c82c34af4903663540e225765) results in a yellow screen when switching to another console and while the following one ([67e3237](https://github.com/freemint/freemint/commit/67e3237e5f0aeba3f3fc441ff73281c1696e78fb)) seemingly fixes it, it is easy to freeze the whole system (my test case was: login on tty1, run `top`, login on tty2, type `ls` + TAB = freeze).
> 
> Note: none of these commits seem to work on Aranym (latest build + latest EmuTOS), the console seems to be frozen. Most likely is not frozen but some screen switching is silently ignored.

 Yes that is the case:  switch to vcons and back works okay, but after a little time in vcons, they freeze.  I am not sure if Aranym problem is just screen switch problem, shouldnt keybpard input still work then? Atleast reset combo, etc?

**Comment by shoggoth77** [comment]
An educated guess - Aranym must be running without fVDI for this to work. The fVDI driver implementation breaks this kind of stuff easily; there is no "real" framebuffer which can be displayed, making it highly difficult for the user to see what's actually happening when you tamper with the "real" Atari screen.

**Comment by skaftetryne** [comment]
It's not a guess - on the Falcon vconsd writes directly to Videl registers when setting up the virtual console framebuffer. I remember hacking it to get a bigger terminal. This is not necessarily a bad thing. I have a graphics card in my main Falcon, and as vconsd used the Videl directly and not the VDI I could have the console and GEM on separate monitors.

**Comment by swe-yc** [comment]
Im running Aranym with NVDI, not fVDI, and it just freeze as soon i try to switch to vconsole. (with 1.16a kernel, which is ok on ataris)

**Comment by mikrosk** [comment]
Yes, fVDI is not an issue here (surprisingly ;)), I have it switched off.

What I'm thinking (and about to test) is that since vcons is agressive enough to be mean to the new re-entrant XBIOS/BIOS handlers, maybe it is EmuTOS which doesnt quite like what vcons is doing. So I'm about to test Aranym + TOS4 + vcons, I think there's a fair chance of it working.

**Comment by swe-yc** [comment]
Mikro: im using Aranym with TOS4... No need to test: same story

**Comment by shoggoth77** [comment]
Well, there are many ways to implement things. @Mikro - exactly how mean is it, what does it do?

**Comment by shoggoth77** [comment]
@skaftetryne (great nick by the way) - I think such behaviour CAN be considered fine, if done right. This is one of the blessings of the platform. Done wrong, e.g. by using Vsetscreen() to re-alloc screen memory and reinitialise the VDI in a user application, or just re-using the logical framebuffer in another screen resolution, it's a disaster. But done right, checking the right cookies, being nice to the system as far as possible, and then access known hardware - that's a good thing. It allows us to do fun stuff that one can just forget about under e.g. Linux.

**Comment by mikrosk** [comment]
> The last working commit with vcons (supplied by Swe, used the vcons.original on my TT) is https://github.com/freemint/freemint/commit/63ac8e3bbb70154fe95e73998cc6309f42e22657. https://github.com/freemint/freemint/commit/50a28d070efc771c82c34af4903663540e225765 results in a yellow screen when switching to another console and while the following one (https://github.com/freemint/freemint/commit/67e3237e5f0aeba3f3fc441ff73281c1696e78fb) seemingly fixes it, it is easy to freeze the whole system (my test case was: login on tty1, run top, login on tty2, type ls + TAB = freeze).

I have just realised an interesting thing. @swe-yc you told me on Discord that your kernel is from May 2006, correct? However those commits which seemingly broken it are from Apr 2004! That means that my bisecting was not correct, Ozk most likely had fixed all the issues in the meantime and then the vcons got broken _again_.

**Comment by mikrosk** [comment]
However there is one other change, related to 1.19.x kernels: "on later 1.19 kernels vcons wont load with "Help!! open new console device failed." message". @mfro0 speculated that it is caused by installing itself as a driver (using Dcntl(DEV_INSTALL)) and failing. Only .xdd's are allowed to do that.

But one thing after another.

**Comment by mikrosk** [comment]
Working again. But beware, vconsd didn't work even with the initial 1.16 kernel with memory protection enabled and 1.19 is no exception here.

Maybe one day I'll convince myself to recompile [vconsd ](http://joy.sophics.cz/www/ttyvdeva.lzh) and investigate more but it's a PITA because the most important component relies on `-mshort` mintlib.

**Comment by swe-yc** [comment]
#response_container_BBPPID{font-family: initial; font-size:initial; color: initial;}  Excellent, thanks a lot!                                                                                                                                                                           From: ***@***.***: 16 June 2026 12:50To: ***@***.*** to: ***@***.***: ***@***.***; ***@***.***: Re: [freemint/freemint] vconsd doesn't work since 1.16.0 beta release (#118)  mikrosk left a comment (freemint/freemint#118)
Working again. But beware, even the initial kernel implementation didn't work with memory protection enabled.
Maybe one day I'll convince myself to recompile vconsd  and investigate more but it's a PITA because the most important component relies on -mshort mintlib.

**Comment by skaftetryne** [comment]
> But beware, vconsd didn't work even with the initial 1.16 kernel with memory protection enabled and 1.19 is no exception here.

vconsd definitely worked with memory protection enabled on older kernels, as this was my exact setup with my Falcon/AB/Nova until I got my Milan in 2001.

**Comment by mikrosk** [comment]
Interesting. The commit I tried (and roughly what @swe-yc is/was using) is from 2004/04/08. And that was already crashing with MP. 

1.15.12 is from 2001/08/01 so yeah, that would match your experience, @skaftetryne. Trouble is that 1.16.x got branched off in 10/2000 (right after 1.15.10 beta release) so that is a looong list of commits to bisect (...and compile).

**Comment by mikrosk** [comment]
Btw @shoggoth77, vconsd doesn't work with SuperVidel. ;) As soon as one tries to switch to another console, everything freezes. 

What is quite interesting because there is no AES involved. But it's true that vconsd is a mixture of direct hardware access and XBIOS calls.

**Comment by skaftetryne** [comment]
Shameless plug: https://atari.joska.no/VidTerm/beta/ works very well with the SuperVidel (actually, it still doesn't work with anything else as both my Falcon/AB/Nova and my Milan has been dismantled for ages), and also has no issues with memory protection. It will let you have 9 terminals on the Videl while the AES is using the SuperVidel.

<img width="1600" height="924" alt="Image" src="https://github.com/user-attachments/assets/ebb3c653-070d-4e02-b5a3-5afddaee5f2a" />

**Comment by mikrosk** [comment]
@skaftetryne I concur with a shameful confession: I still have your email about trying out VIDTerm in CTPCI starred as "todo". :-P

I guess @swe-yc's complain would be that he hasn't got SuperVidel (or at least doesn't have it installed, IIRC?) so a solution with plain Videl is still precious.
